### ISSUE TITLE
Warning reported for [CallerMemberName] argument default value

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#33332]
A warning is reported for the `[CallerMemberName]` argument default value although the argument could be considered non-null.
```C#
#nullable enable

using System.Runtime.CompilerServices;

class Program
{
    static void F([CallerMemberName]string name = null) // warning: converting null literal
    {
    }
}
```
Reported by @onovotny.

**Comment by jaredpar** [comment]
I believe the warning is correct here. Even though the value is _typically_ non-null there are still valid cases where it can be `null`. From the spec: 

> When an optional parameter is annotated with one of the caller info attributes, omitting the corresponding argument in a call does not necessarily cause the default parameter value to be substituted. Instead, if the specified information about the calling context is available, that information will be passed as the argument value.

**Comment by onovotny** [comment]
Is there a case when CallerMemberName would be null (not counting being explicitly passed null)?

**Comment by jaredpar** [comment]
Yes. Consider the following: 

``` 
class C
{
    void F([CallerMemberName]string name = null) 
    {
        dynamic d = new C();
        d.F(); // No caller member info 
    }
}
```

More generally the information will only be provided when the function is explicitly invoked. Implicit invocations do not get caller info (this is from the spec).

**Comment by jaredpar** [comment]
Here are a few samples where it is not actually used. The following code prints out nothing. 

```
class Program
{
    class B
    {
        public B([CallerMemberName]string name = null)
        {
            Console.WriteLine(name);
        }
    }
    interface I
    {
        void F(string name);
    }
    class C : B, I
    {
        public void F([CallerMemberName]string name = null) 
        {
            Console.WriteLine(name);
        }
    }

    static void Main()
    {
        dynamic local = new C();
        local.F();
        I i = local;
        i.F("");
    }
}
```

If it were just one of these cases we could consider special casing but there are too many cases where the information isn't substituted here.

**Comment by onovotny** [comment]
Okay, thanks for the clarification. I see code being littered with pragmas to suppress many of these cases :/
